feat(menu): add HeadlessUI-based menu component

booc0mtaco opened the pull request:
### Summary:

[EDS-739]: Implementation: Popover-like components

- Add in new atomic components for reuse (popover container, popover list item)
- Use Headless UI's dropdown to implement select and menu respectively
- Mark replaced components as deprecated where relevant
- Light refactoring to use popover container in existing popover component

### Test Plan:

- [x] Storybook
- [x] Unit tests (behavior)

[EDS-739]: https://czi-tech.atlassian.net/browse/EDS-739?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

booc0mtaco commented on src/components/Dropdown/Dropdown.module.css:
```diff
@@ -44,7 +44,7 @@
 .dropdown__options {
   @mixin eds-theme-typography-body-text-md;
   /* Allow Dropdown font to shrink in contexts where body font size is smaller. */
-  font-size: inherit; 
```
file contains whitespace fixes picked up by prettier 🤷

booc0mtaco commented:
Not handled in this PR at the moment 

- using popperJS for advanced positioning options (to have parity with `Popover`)
- Implementing a new import of the current `Dropdown` as `Select`

There's also an a11y violation with the value for disabled text on `PopoverLineItem`, so will follow up with design on this one.

dierat commented on src/components/ButtonDropdown/ButtonDropdown.tsx:
```diff
@@ -5,6 +5,7 @@ import styles from './ButtonDropdown.module.css';
 import { DropdownMenu } from '../..';
 import type { ClickableStyleProps } from '../ClickableStyle';
 
+// TODO: This component is deprecated and will be replaced by the Menu component
```
Can you put this deprecated info in the component docstring instead? You can look at the `Banner` component for an example of the current pattern. It has a message at the beginning and a jsdoc tag at the end.
- https://github.com/chanzuckerberg/edu-design-system/blob/main/src/components/Banner/Banner.tsx#L94
- https://github.com/chanzuckerberg/edu-design-system/blob/main/src/components/Banner/Banner.tsx#L113

dierat commented on src/components/ButtonDropdown/ButtonDropdown.stories.tsx:
```diff
@@ -10,6 +11,7 @@ export default {
   title: 'Organisms/Interactive/ButtonDropdown',
   parameters: {
     layout: 'centered',
+    badges: [BADGE.DEPRECATED],
```
Thank you!

dierat commented on src/components/PopoverListItem/__snapshots__/PopoverListItem.test.ts.snap:
```diff
@@ -0,0 +1,73 @@
+`;
+
+exports[`<PopoverListItem /> Disabled story renders snapshot 1`] = `
+<div
+  class="popover-list-item popover-list-item--disabled"
+>
+  <div
+    class="popover-list-item__no-icon"
+  />
+  Disabled
+</div>
```
This is the specific element that's triggering the a11y failure, and I think it's happening because this is a `div` no `aria-disabled` attribute. (It may also need a `role`; I don't know exactly how the test works.)

Disabled interactive elements are exempt from color contrast requirements in WCAG, so it makes sense that this story is failing the test since it's not marked up as disabled.

Source: https://webaim.org/articles/contrast/
>WCAG 2.0 defines four types of "incidental" text that are not required to meet the contrast requirements.
>
>- Inactive: An inactive element, like a disabled Submit button ( Submit ), is identified visually by its lower-contrast state.
>- Pure decoration: Decorative text that is not meant to be read. An example of this might be a picture of a bookshelf on a library homepage. The titles of the books are not meant to be read by the user.
>- Not visible to anyone: Text that is meant to be hidden, like an invisible [skip link](https://webaim.org/techniques/skipnav/) would not need to meet any contrast requirements until it becomes visible.
>- Part of a picture that contains significant other visual content: Text that is not an important part of the information in the image, like a name tag on the shirt of a person in a photo of a party, does not need to meet any contrast requirements.

dierat commented on src/components/Popover/Popover.test.tsx:
```diff
@@ -17,6 +17,7 @@ describe('<Popover />', () => {
   });
 
   it('should open Popover with trigger button', () => {
+    // this test is redundant b/c of the snapshot above
```
That's fair. Should we remove it? Will this be done in a different PR? Is this a TODO comment?

booc0mtaco replied:
yah left a note for myself that this should be removed and left in PR to confirm. Ticket to clean up is https://czi-tech.atlassian.net/browse/EDS-753 and I will trim the comment now :)

dierat commented on src/components/Popover/Popover.stories.tsx:
```diff
@@ -49,6 +49,7 @@ export const Default: StoryObj<Args> = {
 
 export const Arrow: StoryObj<Args> = {
   args: {
+    // maybe gets shallow merged with the default args (?)
```
I'm having trouble understanding the purpose of this comment. Was this a note to yourself?

booc0mtaco replied:
Yah, it's mostly a note that can be removed. @jinlee93 and I were pairing on this and he wasn't sure how the object gets used when consumed will the `args` merge or does the default one get overwritten?

jinlee93 replied:
I believe the `args` are shallow merged, tested via 
![image](https://user-images.githubusercontent.com/86632227/203438393-5463d660-ad7f-4bac-a8d6-1eee3008e354.png)
Which results in `className: 'haw'` and `placement:'top-start'`

dierat commented on src/components/Menu/__snapshots__/Menu.test.tsx.snap:
```diff
@@ -0,0 +1,139 @@
+      </div>
+    </a>
+    <div
+      aria-disabled="true"
```
This element did not trigger the a11y test because it has the expected markup.

dierat commented on src/components/Menu/Menu.module.css:
```diff
@@ -0,0 +1,30 @@
+}
+
+.menu__button {
+  background-color: var(--light-color-button-background);
```
Where's this CSS variable from?

booc0mtaco replied:
that's a GOOD question. I don't know why but some JS files with a bunch of variables popped up for me and I chose one from there. 😕 

The correct one (matches the variable in the figma) is `--eds-theme-color-form-background`

dierat commented:
Is this the figma file for this component? https://www.figma.com/file/Ff7hJoSvXX9hkAmIklnsT2/branch/4xGGJyfPCeoAvYX5Fjssdf/EDS-Paper-v1.1-%7C-UI-Components?node-id=7831%3A24711&t=F8tnXSnksp62mEAr-1

dierat commented on src/components/Menu/Menu.module.css:
```diff
@@ -0,0 +1,30 @@
+
+/* // is getting overridden by a style in clickablestyle */
+.menu__button--icon {
+  color: var(--eds-theme-color-button-secondary-neutral-text);
```
Let's use the token design is using.
```suggestion
  color: var(--eds-theme-color-icon-neutral-default);
```

dierat commented on src/components/Menu/Menu.module.css:
```diff
@@ -0,0 +1,30 @@
+}
+
+.menu__button {
+  background-color: var(--light-color-button-background);
```
```suggestion
  background-color: var(--eds-theme-color-form-background);
```

dierat commented on src/components/Menu/Menu.tsx:
```diff
@@ -0,0 +1,119 @@
+  const buttonClassNames = clsx(styles['menu__button'], className);
+  return (
+    <HeadlessMenu.Button
+      as={ClickableStyle}
```
Have you talked to design about the hover state we're getting from `ClickableStyle` and whether that's what they want?

booc0mtaco replied:
Nah, that is a good call. If you saw the design file, there are some cases and behaviors missing. I am going to add this to the list. Another missing thing is the disabled states (which I left a comment about earlier)

dierat commented:
I dunno how annoying this is to change but, in playing with the menu in storybook, I see that the keyboard focus state for the items in the menu is the light background color change, but I don't think that's enough to indicate focus for keyboard users. The menu has the blue focus outline, but it's going to be hard for users with poorer vision to see which specific item in the list is focused.

In this example, the "MDN Menu" item is currently focused via the keyboard.
![menu component in storybook, open with the MDN Menu list item focused](https://user-images.githubusercontent.com/7761701/203361699-d8cc355b-9388-4780-83fd-5ab0fcd36426.png)

dierat commented on src/components/Menu/Menu.tsx:
```diff
@@ -0,0 +1,119 @@
+  );
+};
+
+/**
```
@jinlee93 are we putting subcomponents in the same file/directory now or still splitting them up?

jinlee93 replied:
I actually wanted to ask you about this.
I am down to start putting subcomponents in the same file/directory as a lot of these subcomponents we really wouldn't be reusing.
However, the file does get really long, stuff could be harder to find, and css class names could get even more complex
@booc0mtaco @dierat

booc0mtaco replied:
My 2¢ on this one is that "bound" sub-components should live in the same folder together. Similar to how headless bundles the parts together. A rule I used in the past was "components get their own folder IFF they can be composed independent of other components".

From what I understand, BM encouraged breaking things down into separate folders, but I think that might be why some of the naming for these popover/dropdowns got wonky...? I chopped some of the reusable bits (`PopoverContainer`, `PopoverListItem`) into separate things and used the common-denominator component as a basis for the name

dierat replied:
>A rule I used in the past was "components get their own folder IFF they can be composed independent of other components".

This sounds right to me. I think it's confusing to have a component in its own folder if it can only be used in conjunction with another specific component—they should be bundled together. I also find it weird that there are directories with no `*.stories.ts/x` file (because subcomponents aren't exposed in storybook).

>However, the file does get really long, stuff could be harder to find, and css class names could get even more complex

I can also see splitting them up into separate files to make it cleaner, but they can still live in the same directory.

dierat replied:
But if we decide to move away from the one-component-per-directory rule, we should update the [code guidelines](https://chanzuckerberg.github.io/edu-design-system/iframe.html?viewMode=story&id=documentation-guidelines--code-guidelines&args=#component-rules).

booc0mtaco replied:
@dierat @jinlee93 yah, I don't have too strong an opinion about how we do the internals of the component, as some cases may benefit from separate subcomponent files more than others. I will turn the guidelines update into a ticket, and we can have a brief discussion at one of our upcoming meetings (or async via comments on that one). How does that sound?

dierat replied:
Sounds good, thanks!

dierat commented on src/components/Menu/Menu.tsx:
```diff
@@ -0,0 +1,119 @@
+  onClick,
+  ...other
+}: MenuItemProps) => {
+  const destinationUrl = !onClick ? href : undefined;
```
Can you explain this line to me? Why not just pass in `href`?

booc0mtaco replied:
Sure, I added this as a behavior preference . if an onClick exists, it will remove the value set for the href so that it doesn't try and do both onClick handling and URL navigation

dierat replied:
Cool! Could you leave a code comment to that effect here?

dierat commented on src/components/PopoverListItem/PopoverListItem.module.css:
```diff
@@ -0,0 +1,32 @@
+
+.popover-list-item--disabled {
+  color: var(--eds-theme-color-text-disabled);
+  pointer-events: none;
```
Is this necessary? Looks like you're not passing in any interactive props into the item if it's disabled, so what pointer events are you removing? `cursor: not-allowed;` might be good though; that's what we use on disabled buttons.

booc0mtaco replied:
good catch. an earlier version of this included an anchor which did have behaviors, and this prevented bubbling up to that to trigger actions.

The cursor is undefined from design side, but I can add that for consistency, and remove this `pointer-events` property.

dierat commented on src/components/PopoverListItem/PopoverListItem.module.css:
```diff
@@ -0,0 +1,32 @@
+
+/* right padding applies space for the icon and the padding for that icon container */
+.popover-list-item__no-icon {
+  padding-right: calc(var(--eds-size-3) + var(--eds-size-1-and-half));
```
Are you adding them together because there's no `4-and-a-half`? I would just pass it in in `rem`s.. We don't use the `size` tokens everywhere. (I'm still not convinced size tokens like these are useful.)

booc0mtaco replied:
I use the two values to show where the quantity comes from. Just above (L26) we use 1.5 for padding, and here the icon size happens to be 3. Lots of components are using various sizes so didn't think they were not part of the existing patterns.

This is fairly complex too. The sizes have a relationship to one another, based on some base size and multipliers ... having tokens for each one means defining different sizes is more work than it has to be IMO. Something to discuss (moving to defining a base grid value and using `calc()` or similar to get other quantities)

dierat replied:
>I use the two values to show where the quantity comes from. Just above (L26) we use 1.5 for padding, and here the icon size happens to be 3.

Ah, cool. Could you move the comment from line 29 down one step so it's above this styling?

dierat commented on src/components/PopoverListItem/PopoverListItem.tsx:
```diff
@@ -0,0 +1,65 @@
+ *
+ * `import {PopoverListItem} from "@chanzuckerberg/eds";`
+ *
+ * This abstracts the structure of a item in a popover, when the popover contains a
```
```suggestion
 * This abstracts the structure of an item in a popover, when the popover contains a
```

dierat commented on src/components/PopoverListItem/PopoverListItem.tsx:
```diff
@@ -0,0 +1,65 @@
+      className,
+    );
+
+    // use `as` to use a/button when appropriate
```
What's this comment referring to? Is it a TODO comment?

booc0mtaco replied:
I can't remember. it was a reminder tho, so chucking it

booc0mtaco commented:
> Is this the figma file for this component? https://www.figma.com/file/Ff7hJoSvXX9hkAmIklnsT2/branch/4xGGJyfPCeoAvYX5Fjssdf/EDS-Paper-v1.1-%7C-UI-Components?node-id=7831%3A24711&t=F8tnXSnksp62mEAr-1

That's the one. Sean is out this week but I think we will have a few more concrete questions for him once he's back.

booc0mtaco commented on src/components/Menu/Menu.module.css:
```diff
@@ -0,0 +1,30 @@
+  background-color: var(--light-color-button-background);
+}
+
+/* // is getting overridden by a style in clickablestyle */
+.menu__button--icon {
```
That's annoying. does it only apply to the things which are using clickable style (since it has `>` rules)? anything else forcing this pattern?

With that in mind, and for this case, it actually appears to need ***three*** of these since the descendant selector 
 and `svg` breaks the tie on the two classes being used on `ClickableStyle` :( [0,3,0] vs [0,2,1] I'll do for now, but worth discussing more (separately)

dierat replied:
>does it only apply to the things which are using clickable style (since it has > rules)?

That's the one I've noticed coming up recently. Maybe we should look into lower the specificity in `ClickableStyle` to make these overrides easier. Though I don't necessarily want it to be easy to override them because I would prefer it if buttons were consistent wherever possible.. 🤔

jinlee93 commented on src/components/Menu/Menu.tsx:
```diff
@@ -0,0 +1,119 @@
+export type MenuItemProps = ExtractProps<typeof HeadlessMenu.Item> & {
+  href?: string;
+  icon?: IconName;
+  onClick?: () => any;
```
```suggestion
  onClick?: MouseEventHandler<HTMLAnchorElement>;
```
would this work?

jinlee93 commented on src/components/Dropdown/Dropdown.module.css:
```diff
@@ -56,7 +56,7 @@
   box-shadow: var(--eds-box-shadow-md);
   /* Ensures the options display over the underlying body elements. */
-  z-index: var(--eds-z-index-100); 
+  z-index: var(--eds-z-index-100);
```
Just an FYI bootstrap has its own set of z-indexes that may clash i.e. [bootstrap 3 in LP](https://github.com/twbs/bootstrap/blob/v3.4.1/less/variables.less#L274-L280) .

Haven't come across that many difficult issues regarding yet

booc0mtaco replied:
I think this file just had whitespace changes in it, so the lines changed should just be that.

jinlee93 commented on src/components/Menu/Menu.tsx:
```diff
@@ -0,0 +1,119 @@
+  <HeadlessMenu.Items
+    as={PopoverContainer}
+    className={clsx(styles['menu__popover'])}
+    data-testid="menu-content"
```
Can we place this `data-testid` attribute on line 27 of `Menu.stories.tsx`?
This way the testid attribute is not publicly exposed (although would that matter?)

booc0mtaco replied:
I have worked in libs where such testIDs got stripped out on publish, i think? I can't remember. Do we have something similar here?

I will move that tho, so we don't have to worry about it

jinlee93 replied:
Good question, not sure if test ids are being stripped out but would be good to have

jinlee93 commented on src/components/PopoverContainer/PopoverContainer.tsx:
```diff
@@ -0,0 +1,39 @@
+  {
+    name: 'offset',
+    options: {
+      offset: [0, 10], // spaces the popover from the trigger element
```
I think the mocks have an offset of 8px

booc0mtaco replied:
This data is a little weird right now. I am not using these values in this PR, but moved this from another file, so all popover containers would have their settings in one place. This does reveal a discrepancy between popovers tho... once they all implement popperJS, they should all use **this** config (this will also prevent circular dependencies) and we would change the values to 8 if that's correct for everyone.

jinlee93 replied:
Makes sense, thanks

jinlee93 commented on src/components/PopoverListItem/PopoverListItem.stories.tsx:
```diff
@@ -0,0 +1,39 @@
+
+export const Default: StoryObj<Args> = {
+  args: {
+    children: <>Default list item</>,
```
nit:
```suggestion
    children: 'Default list item'
```

booc0mtaco replied:
ah i didn't know it would allow strings like that; will update

jinlee93 commented on src/components/PopoverListItem/PopoverListItem.tsx:
```diff
@@ -0,0 +1,65 @@
+   * Whether the list item is treated as highlighted in its container
+   */
+  active?: boolean;
+  children: ReactNode;
```
```suggestion
  /**
   * Child node(s) that can be nested inside component
   */
  children: ReactNode;
```

booc0mtaco commented on src/components/Menu/Menu.tsx:
```diff
@@ -0,0 +1,119 @@
+ * be used when there is a discrete number of actions such as user navigations (e.g.,
+ * a profile menu with links to settings) or a set of actions
+ */
+export const Menu = (props: MenuProps) => {
```
In our xfn meeting we landed on using Menu/Select for the two components to avoid ambiguity. All three of these things sorta drop down, and we already have other (deprecated) components using "Dropdown" in the name, so trying to avoid that in the component itself.

The figma currently has conflicting naming too, so will take a note for that to get tidied up

dierat replied:
Thanks!

booc0mtaco commented on src/components/PopoverContent/PopoverContent.tsx:
```diff
@@ -88,7 +84,7 @@ export const PopoverContent = ({
     >
-      <div className={bodyComponentClassName}>{children}</div>
+      <PopoverContainer className={bodyClassName}>{children}</PopoverContainer>
       {showArrow && (
```
haha, I will add it to the list. I think for menu / select we don't want arrow according to [figma](https://www.figma.com/file/Ff7hJoSvXX9hkAmIklnsT2/branch/4xGGJyfPCeoAvYX5Fjssdf/EDS-Paper-v1.1-%7C-UI-Components?node-id=7831%3A24711&t=WIALim2FJJoByEXT-0) but it is undefined at the moment which popover-like things should support arrows

booc0mtaco commented:
> I dunno how annoying this is to change but, in playing with the menu in storybook, I see that the keyboard focus state for the items in the menu is the light background color change, but I don't think that's enough to indicate focus for keyboard users. The menu has the blue focus outline, but it's going to be hard for users with poorer vision to see which specific item in the list is focused.
> 
> In this example, the "MDN Menu" item is currently focused via the keyboard. ![menu component in storybook, open with the MDN Menu list item focused](https://user-images.githubusercontent.com/7761701/203361699-d8cc355b-9388-4780-83fd-5ab0fcd36426.png)

@dierat Gonna add this to the list, as I did follow up with sean if we should have focus ring and it sounded like that wasn't needed. it

booc0mtaco commented on src/components/DropdownMenu/DropdownMenu.tsx:
```diff
@@ -18,6 +18,7 @@ import {
   TAB_KEYCODE,
 } from '../../util/keycodes';
 
+// This component is deprecated and will be replaced by the Menu component
```
sounds good. also another supporting fact for grouping things together since marking subcomponents as deprecated gets complicated if they are all separated out.

booc0mtaco commented:
* Added https://czi-tech.atlassian.net/browse/EDS-754 to handle adding the `Select` export from EDS
* Added https://czi-tech.atlassian.net/browse/EDS-755 to handle doing popperJS (re-)factoring

dierat commented:
> > I dunno how annoying this is to change but, in playing with the menu in storybook, I see that the keyboard focus state for the items in the menu is the light background color change, but I don't think that's enough to indicate focus for keyboard users. The menu has the blue focus outline, but it's going to be hard for users with poorer vision to see which specific item in the list is focused.
> > In this example, the "MDN Menu" item is currently focused via the keyboard. ![menu component in storybook, open with the MDN Menu list item focused](https://user-images.githubusercontent.com/7761701/203361699-d8cc355b-9388-4780-83fd-5ab0fcd36426.png)
> 
> @dierat Gonna add this to the list, as I did follow up with sean if we should have focus ring and it sounded like that wasn't needed. it

We need it. This is just going to come back as an accessibility issue via Lodestone when someone uses it in the product if we don't add a high contrast focus ring to the individual list items for keyboard users.

jinlee93 reviewed (approved):
Thanks for all the Dropdown work and clarification!

booc0mtaco commented:
Added final ticket to track some questions that came up here https://czi-tech.atlassian.net/browse/EDS-759
